Update Jest examples and docs (#31633)
vercel/next.js: docs/testing.md

```bash
npm install --save-dev jest @testing-library/react @testing-library/jest-dom react-test-renderer
```

Create a `jest.config.js` file in your project's root directory and add the following:

```jsx
// jest.config.js
const nextJest = require('next/jest')

const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
  dir: './',
})

// Add any custom config to be passed to Jest
const customJestConfig = {
  setupFilesAfterEnv: ['<rootDir>/jest.setup.js'],
}

// createJestConfig is exported this way to ensure that next/jest can load the Next.js config which is async
module.exports = createJestConfig(customJestConfig)
```

Under the hood, `next/jest` is automatically configuring Jest for you, including:

- Setting up `transform` using [SWC](https://nextjs.org/docs/advanced-features/compiler)
- Auto mocking stylesheets (`.css`, `.module.css`, and their scss variants) and image imports
- Loading `.env` (and all variants) into `process.env`
- Ignoring `node_modules` from test resolving and transforms
- Ignoring `.next` from test resolving

> - Loading `next.config.js` for flags that enable experimental SWC transforms

Reviewer: ```suggestion
- Loading `next.config.js` for flags that enable SWC transforms
```

### Setting up Jest (with Babel)

> If you opt out of the [Rust Compiler](https://nextjs.org/docs/advanced-features/compiler), you will need to manually configure Jest and install `babel-jest` and `identity-obj-proxy` in addition to the packages above.

Reviewer: ```suggestion
If you opt-out of the [Rust Compiler](https://nextjs.org/docs/advanced-features/compiler), you will need to manually configure Jest and install `babel-jest` and `identity-obj-proxy` in addition to the packages above.
```

> ### Quickstart

Reviewer 2: I wonder if I'm being pedantic here, what are your thoughts on converting this into steps, using the same heading titles?

Author: Could you expand on what you mean here? 😊

Reviewer 2: Sorry. I just mean, would It be clearer for us to explain the QuickStart as a set of steps:

```
## Step 1: install

## Step2: do something else
```

This may not be the case here. I don't have a hard and fast rule on this, I just think it makes skimming content easier, you can drop in a a step if need etc.. I don't think this fits in all cases, was wondering what your thoughts are with this section.

I quite like how you have done it tbh, just raising the question. ☺️

Author: Ah, thank you! I agree with you, it's good to give people context at the start. As you said, I'm not sure how the steps would fit this case, but I've added a [quick summary here](https://github.com/vercel/next.js/pull/31633/commits/3968e9afe0e2a5d3a22d1ad42d013224f12deca6). wdyt?

You can use `create-next-app` with the [with-jest](https://github.com/vercel/next.js/tree/canary/examples/with-jest) example to quickly get started with Jest and React Testing Library:

```bash
npx create-next-app@latest --example with-jest with-jest-app
```

### Setting up Jest (with SWC)

> Since the release of (Next.js 12)[https://nextjs.org/blog/next-12], Next.js now has built-in configuration for Jest. 

Reviewer: Should we say 12.1 or 12.0.[x]?

Reviewer 3: ```suggestion
Since the release of [Next.js 12](https://nextjs.org/blog/next-12), Next.js now has built-in configuration for Jest. 
```

> To set up Jest, install `jest` , `@testing-library/react`, `@testing-library/jest-dom` and `react-test-renderer`.

Reviewer 2: ```suggestion
To set up Jest, install `jest` , `@testing-library/react`, `@testing-library/jest-dom` and `react-test-renderer`:
```

```bash
npm install --save-dev jest @testing-library/react @testing-library/jest-dom react-test-renderer
```

Create a `jest.config.js` file in your project's root directory and add the following:

```jsx
// jest.config.js

> const nextJest = require('next/jest')

Reviewer 2: Love the code comments here. Adds some context for the reader 🙌

const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
  dir: './',
})

// Add any custom config to be passed to Jest
const customJestConfig = {
  setupFilesAfterEnv: ['<rootDir>/jest.setup.js'],
}

// createJestConfig is exported this way to ensure that next/jest can load the Next.js config which is async
module.exports = createJestConfig(customJestConfig)
```

Under the hood, `next/jest` is automatically configuring Jest for you, including:

> - Auto mocking stylesheets (.css, .module.css, and their scss variants) and image imports

Reviewer: ```suggestion
- Auto mocking stylesheets (`.css`, `.module.css`, and their scss variants) and image imports
```

> - Setting up `transform` using SWC

Reviewer: Could we link SWC to https://github.com/vercel/next.js/pull/31485?

Reviewer: Would make this dependent on that PR but I think it's okay

Author: Yes, I think we could link to the docs here.

> - Loading .env (and all variants) into process.env
> - Ignoring node_modules from test resolving and transforms
> - Ignoring .next from test resolving
> - Loading next.config.js for flags that enable experimental SWC transforms

Reviewer: ```suggestion
- Loading `.env` (and all variants) into `process.env`
- Ignoring `node_modules` from test resolving and transforms
- Ignoring `.next` from test resolving
- Loading `next.config.js` for flags that enable experimental SWC transforms
```

### Setting up Jest (with Babel)

> If you opt out of SWC, you will need to manually configure Jest and install `babel-jest` and `identity-obj-proxy` in addition to the packages above.

Reviewer: Same here. Replace SWC with Rust Compiler and link to the docs?

> If you opt-out of SWC and use Babel instead. In addition to packages above, you will need to install the `babel-jest` and `identity-obj-proxy` packages, and manually configure Jest.

Reviewer 4: Recommend using "opt out" instead of "opt-out" based on the context. Also, "If you opt-out of SWC and use Babel instead." is an incomplete sentence – did you mean to add something?

Author: Hm, yes, I think you're right. That doesn't read smoothly, does it?

Reviewer 4: Maybe, "If you opt out of SWC, use Babel instead" would work better? I don't think that "and" is supposed to be there :)

Author: How about? 

> If you opt out of SWC, you will need to manually configure Jest and install `babel-jest` and `identity-obj-proxy` in addition to the packages above.

Reviewer 4: Perfect!

Author: ty!

}
```

You can learn more about each configuration option in the [Jest docs](https://jestjs.io/docs/configuration).

**Handling stylesheets and image imports**

> Styleheets and images aren't used in the tests but importing them may cause errors, so we will need to mock them. Create the mock files we referenced in the configuration above - `fileMock.js` and `styleMock.js` - inside a `__mocks__` directory:

Reviewer 2: ```suggestion
Styleheets and images aren't used in the tests but importing them may cause errors, so they will need to be mocked. Create the mock files referenced in the configuration above - `fileMock.js` and `styleMock.js` - inside a `__mocks__` directory:
```

> ```json

Reviewer 3: ```suggestion
```js
```

Author: Thank you for spotting that @stefanprobst 🙏🏼

// __mocks__/fileMock.js
(module.exports = "test-file-stub")
```

> ```json

Reviewer 3: ```suggestion
```js
```

// __mocks__/styleMock.js
module.exports = {};
```

If you're running into the issue `"Failed to parse src "test-file-stub" on 'next/image'"`, add a '/' to your fileMock.

> ```json

Reviewer 3: ```suggestion
```js
```

// __mocks__/fileMock.js
(module.exports = "/test-file-stub")
```

For more information on handling static assets, please refer to the [Jest Docs](https://jestjs.io/docs/webpack#handling-static-assets).

**Optional: Extend Jest with custom matchers**

`@testing-library/jest-dom` includes a set of convenient [custom matchers](https://github.com/testing-library/jest-dom#custom-matchers) such as `.toBeInTheDocument()` making it easier to write tests. You can import the custom matchers for every test by adding the following option to the Jest configuration file:

> ```json

Reviewer 3: ```suggestion
```js
```
